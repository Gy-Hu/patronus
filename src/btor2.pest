// Copyright 2023 The Regents of the University of California
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@berkeley.edu>

WHITESPACE = _{ " " | "\t" | "\r" }

btor2File = {SOI ~ lines ~ EOI}
lines = { ((line | WHITESPACE*) ~ NEWLINE)* }

// tokens
number = @{ ASCII_DIGIT+ }
unaryOp = @{ "not" | "inc" | "dec" | "neg" | "redand" | "redor" | "redxor" }
binaryOp = @{ "iff" | "implies" | "sgt" | "ugt" | "sgte" | "ugte" | "slt" | "ult" | "slte" | "ulte" | "and" | "nand" |
              "nor" | "or" | "xnor" | "xor" | "rol" | "ror" | "sll" | "sra" | "srl" | "add" | "mul" | "sdiv" | "udiv" |
              "smod" | "srem" | "urem" | "sub" | "saddo" | "uaddo" | "sdivo" | "udivo" | "smulo" | "umulo" | "ssubo" |
              "usubo" | "concat" | "eq" }

// each line is a declaration that is preceded by a unique ID
line = { number ~ (
declBvSort | declArraySort |
constZero | constOne | constBin | constDec | constHex | constOnes |
declState | declStateInit | declStateNext |
declBinOp |
declBad
)}

// individual commands without their ID
declBvSort = { "sort" ~ "bitvec" ~ number  }
declArraySort = { "sort" ~ "array" ~ number ~ number }
constZero = { "zero" ~ number }
constOne = { "one"  ~ number }
constBin = { "const" ~ number ~ ASCII_BIN_DIGIT* }
constDec = { "constd" ~ number ~ ASCII_DIGIT* }
constHex = { "consth" ~ number ~ ASCII_HEX_DIGIT* }
constOnes = { "ones" ~ number }
declState = { "state" ~ number  }
declStateInit = { "init" ~ number ~ number ~ number }
declStateNext = { "next" ~ number ~ number ~ number }
declBinOp = { binaryOp ~ number ~ number ~ number }
declBad = { "bad" ~ number }