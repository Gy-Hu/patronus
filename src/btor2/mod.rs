// Copyright 2023 The Regents of the University of California
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@berkeley.edu>
mod parse;
mod serialize;
mod witness;

pub use parse::{parse_file, parse_str};
pub use serialize::{serialize, serialize_to_str};
pub use witness::{parse_witness, parse_witnesses, print_witness, witness_to_string};
