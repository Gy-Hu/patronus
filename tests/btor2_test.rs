// Copyright 2023 The Regents of the University of California
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@berkeley.edu>

use libpatron::btor2;
use libpatron::ir::Context;
use libpatron::ir::SerializableIrNode;

const COUNT_2: &str = r#"
1 sort bitvec 3
2 zero 1
3 state 1
4 init 1 3 2
5 one 1
6 add 1 3 5
7 next 1 3 6
8 ones 1
9 sort bitvec 1
10 eq 9 3 8
11 bad 10
"#;

#[test]
fn parse_count2() {
    let mut ctx = Context::default();
    let sys = btor2::parse_str(&mut ctx, COUNT_2, Some("count2")).unwrap();
    println!("{}", sys.serialize_to_str(&ctx));
}
