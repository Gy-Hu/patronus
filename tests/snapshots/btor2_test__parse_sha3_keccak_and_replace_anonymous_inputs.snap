---
source: tests/btor2_test.rs
expression: sys.serialize_to_str(&ctx)
---
sha3_keccak.w2.replace_variables
input byte_num : bv<2>
input is_last : bv<1>
input in_ready : bv<1>
input in : bv<32>
input reset : bv<1>
input clk : bv<1>
output out_ready_0 : bv<1> = i[22]
output %1788 : bv<512> = concat(f_out[1543:1536], concat(f_out[1551:1544], concat(f_out[1559:1552], concat(f_out[1567:1560], concat(f_out[1575:1568], concat(f_out[1583:1576], concat(f_out[1591:1584], concat(f_out[1599:1592], concat(f_out[1479:1472], concat(f_out[1487:1480], concat(f_out[1495:1488], concat(f_out[1503:1496], concat(f_out[1511:1504], concat(f_out[1519:1512], concat(f_out[1527:1520], concat(f_out[1535:1528], concat(f_out[1415:1408], concat(f_out[1423:1416], concat(f_out[1431:1424], concat(f_out[1439:1432], concat(f_out[1447:1440], concat(f_out[1455:1448], concat(f_out[1463:1456], concat(f_out[1471:1464], concat(f_out[1351:1344], concat(f_out[1359:1352], concat(f_out[1367:1360], concat(f_out[1375:1368], concat(f_out[1383:1376], concat(f_out[1391:1384], concat(f_out[1399:1392], concat(f_out[1407:1400], concat(f_out[1287:1280], concat(f_out[1295:1288], concat(f_out[1303:1296], concat(f_out[1311:1304], concat(f_out[1319:1312], concat(f_out[1327:1320], concat(f_out[1335:1328], concat(f_out[1343:1336], concat(f_out[1223:1216], concat(f_out[1231:1224], concat(f_out[1239:1232], concat(f_out[1247:1240], concat(f_out[1255:1248], concat(f_out[1263:1256], concat(f_out[1271:1264], concat(f_out[1279:1272], concat(f_out[1159:1152], concat(f_out[1167:1160], concat(f_out[1175:1168], concat(f_out[1183:1176], concat(f_out[1191:1184], concat(f_out[1199:1192], concat(f_out[1207:1200], concat(f_out[1215:1208], concat(f_out[1095:1088], concat(f_out[1103:1096], concat(f_out[1111:1104], concat(f_out[1119:1112], concat(f_out[1127:1120], concat(f_out[1135:1128], concat(f_out[1143:1136], f_out[1151:1144])))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))
output 0 : bv<1> = 1'b0
node %1859 : bv<1> = ite(__synth_change_variable_19, ite(__synth_variable_135, 0, ite(__synth_variable_134, 0, ite(__synth_variable_133, state, ite(__synth_variable_132, out_ready_0, ite(__synth_variable_131, 0, ite(__synth_variable_130, is_last, ite(__synth_variable_129, in_ready, ite(__synth_variable_128, reset, clk)))))))), 0)
node %1890 : bv<1600> = ite(%1859, concat(xor(ite(__synth_change_variable_17, 576'x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, 576'x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000), f_out[1599:1024]), f_out[1023:0]), f_out)
node %1891 : bv<64> = %1890[63:0]
node %1892 : bv<64> = %1890[383:320]
node %1893 : bv<64> = %1890[703:640]
node %1894 : bv<64> = %1890[1023:960]
node %1895 : bv<64> = %1890[1343:1280]
node %1899 : bv<64> = xor(xor(xor(xor(%1895, %1894), %1893), %1892), %1891)
node %1902 : bv<64> = concat(%1899[62:0], %1899[63])
node %1903 : bv<64> = %1890[191:128]
node %1904 : bv<64> = %1890[511:448]
node %1905 : bv<64> = %1890[831:768]
node %1906 : bv<64> = %1890[1151:1088]
node %1907 : bv<64> = %1890[1471:1408]
node %1911 : bv<64> = xor(xor(xor(xor(%1907, %1906), %1905), %1904), %1903)
node %1912 : bv<64> = %1890[1087:1024]
node %1914 : bv<64> = xor(xor(%1912, %1911), %1902)
node %1917 : bv<64> = concat(%1914[8:0], %1914[63:9])
node %1918 : bv<64> = %1890[127:64]
node %1919 : bv<64> = %1890[447:384]
node %1920 : bv<64> = %1890[767:704]
node %1921 : bv<64> = %1890[1407:1344]
node %1925 : bv<64> = xor(xor(xor(xor(%1921, %1912), %1920), %1919), %1918)
node %1928 : bv<64> = concat(%1925[62:0], %1925[63])
node %1929 : bv<64> = %1890[255:192]
node %1930 : bv<64> = %1890[575:512]
node %1931 : bv<64> = %1890[895:832]
node %1932 : bv<64> = %1890[1215:1152]
node %1933 : bv<64> = %1890[1535:1472]
node %1937 : bv<64> = xor(xor(xor(xor(%1933, %1932), %1931), %1930), %1929)
node %1939 : bv<64> = xor(xor(%1907, %1937), %1928)
node %1942 : bv<64> = concat(%1939[1:0], %1939[63:2])
node %1947 : bv<64> = concat(%1911[62:0], %1911[63])
node %1948 : bv<64> = %1890[319:256]
node %1949 : bv<64> = %1890[639:576]
node %1950 : bv<64> = %1890[959:896]
node %1951 : bv<64> = %1890[1279:1216]
node %1952 : bv<64> = %1890[1599:1536]
node %1956 : bv<64> = xor(xor(xor(xor(%1952, %1951), %1950), %1949), %1948)
node %1958 : bv<64> = xor(xor(%1929, %1956), %1947)
node %1961 : bv<64> = concat(%1958[61:0], %1958[63:62])
node %1967 : bv<64> = concat(%1937[62:0], %1937[63])
node %1969 : bv<64> = xor(xor(%1949, %1899), %1967)
node %1972 : bv<64> = concat(%1969[22:0], %1969[63:23])
node %1979 : bv<64> = concat(%1956[62:0], %1956[63])
node %1981 : bv<64> = xor(xor(%1893, %1925), %1979)
node %1984 : bv<64> = concat(%1981[24:0], %1981[63:25])
node %1996 : bv<64> = xor(xor(%1951, %1899), %1967)
node %1999 : bv<64> = concat(%1996[27:0], %1996[63:28])
node %2001 : bv<64> = xor(xor(%1895, %1925), %1979)
node %2004 : bv<64> = concat(%2001[36:0], %2001[63:37])
node %2008 : bv<64> = xor(xor(%1918, %1911), %1902)
node %2011 : bv<64> = concat(%2008[7:0], %2008[63:8])
node %2017 : bv<64> = xor(xor(%1904, %1937), %1928)
node %2020 : bv<64> = concat(%2017[48:0], %2017[63:49])
node %2026 : bv<64> = xor(xor(%1931, %1956), %1947)
node %2029 : bv<64> = concat(%2026[53:0], %2026[63:54])
node %2041 : bv<64> = xor(xor(%1906, %1937), %1928)
node %2044 : bv<64> = concat(%2041[57:0], %2041[63:58])
node %2046 : bv<64> = xor(xor(%1933, %1956), %1947)
node %2049 : bv<64> = concat(%2046[62:0], %2046[63])
node %2053 : bv<64> = xor(xor(%1948, %1899), %1967)
node %2056 : bv<64> = concat(%2053[45:0], %2053[63:46])
node %2062 : bv<64> = xor(xor(%1892, %1925), %1979)
node %2065 : bv<64> = concat(%2062[55:0], %2062[63:56])
node %2071 : bv<64> = xor(xor(%1920, %1911), %1902)
node %2074 : bv<64> = concat(%2071[38:0], %2071[63:39])
node %2086 : bv<64> = xor(xor(%1894, %1925), %1979)
node %2089 : bv<64> = concat(%2086[43:0], %2086[63:44])
node %2091 : bv<64> = xor(xor(%1921, %1911), %1902)
node %2094 : bv<64> = concat(%2091[35:0], %2091[63:36])
node %2098 : bv<64> = xor(xor(%1903, %1937), %1928)
node %2101 : bv<64> = concat(%2098[2:0], %2098[63:3])
node %2107 : bv<64> = xor(xor(%1930, %1956), %1947)
node %2110 : bv<64> = concat(%2107[18:0], %2107[63:19])
node %2116 : bv<64> = xor(xor(%1950, %1899), %1967)
node %2119 : bv<64> = concat(%2116[60:0], %2116[63:61])
node %2131 : bv<64> = xor(xor(%1932, %1956), %1947)
node %2134 : bv<64> = concat(%2131[19:0], %2131[63:20])
node %2136 : bv<64> = xor(xor(%1952, %1899), %1967)
node %2140 : bv<64> = xor(xor(%1891, %1925), %1979)
node %2143 : bv<64> = concat(%2140[49:0], %2140[63:50])
node %2149 : bv<64> = xor(xor(%1919, %1911), %1902)
node %2152 : bv<64> = concat(%2149[42:0], %2149[63:43])
node %2158 : bv<64> = xor(xor(%1905, %1937), %1928)
node %2161 : bv<64> = concat(%2158[20:0], %2158[63:21])
node %521 : bv<1> = f_permutation_.i[21]
node %519 : bv<1> = f_permutation_.i[19]
node %517 : bv<1> = f_permutation_.i[14]
node %515 : bv<1> = f_permutation_.i[13]
node %513 : bv<1> = f_permutation_.i[12]
node %511 : bv<1> = f_permutation_.i[11]
node %509 : bv<1> = f_permutation_.i[9]
node %507 : bv<1> = f_permutation_.i[6]
node %505 : bv<1> = f_permutation_.i[5]
node %503 : bv<1> = f_permutation_.i[4]
node %501 : bv<1> = f_permutation_.i[3]
node %2181 : bv<64> = xor(%2136, and(not(%2134), %2161))
node %540 : bv<1> = f_permutation_.i[18]
node %538 : bv<1> = f_permutation_.i[17]
node %536 : bv<1> = f_permutation_.i[15]
node %531 : bv<1> = f_permutation_.i[10]
node %529 : bv<1> = f_permutation_.i[7]
node %526 : bv<1> = f_permutation_.i[1]
node %527 : bv<1> = or(f_permutation_.i[0], %526)
node %528 : bv<1> = or(%527, %501)
node %556 : bv<1> = f_permutation_.i[22]
node %547 : bv<1> = f_permutation_.i[8]
node %569 : bv<1> = f_permutation_.i[20]
node %566 : bv<1> = f_permutation_.i[16]
node %573 : bv<1> = f_permutation_.i[2]
node %1867 : bv<1> = ite(__synth_change_variable_16, ite(__synth_variable_119, 0, ite(__synth_variable_118, 0, ite(__synth_variable_117, 0, ite(__synth_variable_116, state, ite(__synth_variable_115, out_ready_0, ite(__synth_variable_114, 0, ite(__synth_variable_113, is_last, ite(__synth_variable_112, in_ready, clk)))))))), reset)
node %1840 : bv<1> = ite(__synth_change_variable_7, ite(__synth_variable_62, 0, ite(__synth_variable_61, 0, ite(__synth_variable_60, 0, ite(__synth_variable_59, state, ite(__synth_variable_58, out_ready_0, ite(__synth_variable_57, 0, ite(__synth_variable_56, is_last, ite(__synth_variable_55, in_ready, clk)))))))), reset)
node %1818 : bv<1> = ite(__synth_change_variable_11, ite(__synth_variable_87, 0, ite(__synth_variable_86, 0, ite(__synth_variable_85, 0, ite(__synth_variable_84, state, ite(__synth_variable_83, out_ready_0, ite(__synth_variable_82, is_last, ite(__synth_variable_81, in_ready, ite(__synth_variable_80, reset, clk)))))))), 0)
node %1795 : bv<1> = ite(__synth_change_variable_14, ite(__synth_variable_103, 0, ite(__synth_variable_102, 0, ite(__synth_variable_101, state, ite(__synth_variable_100, out_ready_0, ite(__synth_variable_99, 0, ite(__synth_variable_98, is_last, ite(__synth_variable_97, in_ready, ite(__synth_variable_96, reset, clk)))))))), 0)
node %1796 : bv<1> = not(%1795)
node %1819 : bv<1> = not(%1818)
state i : bv<23>
  [next] ite(ite(__synth_change_variable_0, ite(__synth_variable_7, 0, ite(__synth_variable_6, 0, ite(__synth_variable_5, 0, ite(__synth_variable_4, state, ite(__synth_variable_3, out_ready_0, ite(__synth_variable_2, 0, ite(__synth_variable_1, is_last, ite(__synth_variable_0, in_ready, clk)))))))), reset), 23'x000000, concat(i[21:0], and(ite(__synth_change_variable_1, ite(__synth_variable_15, 0, ite(__synth_variable_14, 0, ite(__synth_variable_13, 0, ite(__synth_variable_12, out_ready_0, ite(__synth_variable_11, 0, ite(__synth_variable_10, is_last, ite(__synth_variable_9, in_ready, ite(__synth_variable_8, reset, clk)))))))), state), ite(__synth_change_variable_2, ite(__synth_variable_23, 0, ite(__synth_variable_22, 0, ite(__synth_variable_21, state, ite(__synth_variable_20, out_ready_0, ite(__synth_variable_19, 0, ite(__synth_variable_18, is_last, ite(__synth_variable_17, in_ready, ite(__synth_variable_16, reset, clk)))))))), 0))))
state f_out : bv<1600>
  [next] ite(%1867, 1600'x0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, ite(or(f_permutation_.calc, %1859), concat(xor(%2181[63], or(or(or(or(or(or(or(or(or(or(or(or(%526, %573), %505), %507), %513), %515), %517), %536), %566), %540), %519), %569), %556)), concat(%2181[62:32], concat(xor(%2181[31], or(or(or(or(or(or(or(or(or(%573, %503), %505), %509), %531), %511), %540), %519), %521), %556)), concat(%2181[30:16], concat(xor(%2181[15], or(or(or(or(or(or(or(or(or(or(or(or(or(%527, %573), %501), %505), %507), %509), %511), %515), %517), %536), %538), %519), %569), %556)), concat(%2181[14:8], concat(xor(%2181[7], or(or(or(or(or(or(or(or(or(%528, %505), %529), %547), %511), %513), %515), %566), %519), %569)), concat(%2181[6:4], concat(xor(%2181[3], or(or(or(or(or(or(or(or(or(or(or(or(%526, %501), %507), %529), %547), %509), %531), %511), %513), %515), %538), %540), %556)), concat(%2181[2], concat(xor(%2181[1], or(or(or(or(or(or(or(or(%528, %529), %531), %511), %513), %517), %536), %538), %540)), concat(xor(%2181[0], or(or(or(or(or(or(or(or(or(or(or(%1859, %501), %503), %505), %507), %509), %511), %513), %515), %517), %519), %521)), concat(xor(%2134, and(not(%2161), %2152)), concat(xor(%2161, and(not(%2152), %2143)), concat(xor(%2152, and(not(%2143), %2136)), concat(xor(%2143, and(not(%2136), %2134)), concat(xor(%2094, and(not(%2089), %2119)), concat(xor(%2089, and(not(%2119), %2110)), concat(xor(%2119, and(not(%2110), %2101)), concat(xor(%2110, and(not(%2101), %2094)), concat(xor(%2101, and(not(%2094), %2089)), concat(xor(%2049, and(not(%2044), %2074)), concat(xor(%2044, and(not(%2074), %2065)), concat(xor(%2074, and(not(%2065), %2056)), concat(xor(%2065, and(not(%2056), %2049)), concat(xor(%2056, and(not(%2049), %2044)), concat(xor(%2004, and(not(%1999), %2029)), concat(xor(%1999, and(not(%2029), %2020)), concat(xor(%2029, and(not(%2020), %2011)), concat(xor(%2020, and(not(%2011), %2004)), concat(xor(%2011, and(not(%2004), %1999)), concat(xor(%1942, and(not(%1917), %1984)), concat(xor(%1917, and(not(%1984), %1972)), concat(xor(%1984, and(not(%1972), %1961)), concat(xor(%1972, and(not(%1961), %1942)), xor(%1961, and(not(%1942), %1917))))))))))))))))))))))))))))))))))))), f_out))
state __synth_variable_104 : bv<1>
  [next] __synth_variable_104
state __synth_variable_105 : bv<1>
  [next] __synth_variable_105
state __synth_variable_106 : bv<1>
  [next] __synth_variable_106
state __synth_variable_107 : bv<1>
  [next] __synth_variable_107
state state : bv<1>
  [next] ite(ite(__synth_change_variable_3, ite(__synth_variable_31, 0, ite(__synth_variable_30, 0, ite(__synth_variable_29, 0, ite(__synth_variable_28, state, ite(__synth_variable_27, out_ready_0, ite(__synth_variable_26, 0, ite(__synth_variable_25, is_last, ite(__synth_variable_24, in_ready, clk)))))))), reset), 0, ite(ite(__synth_change_variable_4, ite(__synth_variable_39, 0, ite(__synth_variable_38, 0, ite(__synth_variable_37, 0, ite(__synth_variable_36, state, ite(__synth_variable_35, out_ready_0, ite(__synth_variable_34, 0, ite(__synth_variable_33, in_ready, ite(__synth_variable_32, reset, clk)))))))), is_last), 1'b1, state))
state __synth_variable_108 : bv<1>
  [next] __synth_variable_108
state __synth_variable_109 : bv<1>
  [next] __synth_variable_109
state __synth_variable_110 : bv<1>
  [next] __synth_variable_110
state __synth_variable_111 : bv<1>
  [next] __synth_variable_111
state __synth_change_variable_15 : bv<1>
  [next] __synth_change_variable_15
state __synth_variable_112 : bv<1>
  [next] __synth_variable_112
state __synth_variable_113 : bv<1>
  [next] __synth_variable_113
state __synth_variable_114 : bv<1>
  [next] __synth_variable_114
state __synth_variable_115 : bv<1>
  [next] __synth_variable_115
state __synth_variable_116 : bv<1>
  [next] __synth_variable_116
state __synth_variable_117 : bv<1>
  [next] __synth_variable_117
state __synth_variable_118 : bv<1>
  [next] __synth_variable_118
state __synth_variable_119 : bv<1>
  [next] __synth_variable_119
state __synth_change_variable_16 : bv<1>
  [next] __synth_change_variable_16
state __synth_change_variable_17 : bv<1>
  [next] __synth_change_variable_17
state __synth_variable_120 : bv<1>
  [next] __synth_variable_120
state __synth_variable_121 : bv<1>
  [next] __synth_variable_121
state __synth_variable_122 : bv<1>
  [next] __synth_variable_122
state __synth_variable_123 : bv<1>
  [next] __synth_variable_123
state __synth_variable_124 : bv<1>
  [next] __synth_variable_124
state __synth_variable_125 : bv<1>
  [next] __synth_variable_125
state __synth_variable_126 : bv<1>
  [next] __synth_variable_126
state __synth_variable_127 : bv<1>
  [next] __synth_variable_127
state __synth_change_variable_18 : bv<1>
  [next] __synth_change_variable_18
state __synth_variable_128 : bv<1>
  [next] __synth_variable_128
state __synth_variable_129 : bv<1>
  [next] __synth_variable_129
state __synth_variable_130 : bv<1>
  [next] __synth_variable_130
state __synth_variable_131 : bv<1>
  [next] __synth_variable_131
state __synth_variable_132 : bv<1>
  [next] __synth_variable_132
state __synth_variable_133 : bv<1>
  [next] __synth_variable_133
state __synth_variable_134 : bv<1>
  [next] __synth_variable_134
state __synth_variable_135 : bv<1>
  [next] __synth_variable_135
state __synth_change_variable_19 : bv<1>
  [next] __synth_change_variable_19
state __synth_variable_136 : bv<1>
  [next] __synth_variable_136
state __synth_variable_137 : bv<1>
  [next] __synth_variable_137
state __synth_variable_138 : bv<1>
  [next] __synth_variable_138
state __synth_variable_139 : bv<1>
  [next] __synth_variable_139
state __synth_variable_140 : bv<1>
  [next] __synth_variable_140
state __synth_variable_141 : bv<1>
  [next] __synth_variable_141
state __synth_variable_142 : bv<1>
  [next] __synth_variable_142
state __synth_variable_143 : bv<1>
  [next] __synth_variable_143
state __synth_change_variable_20 : bv<1>
  [next] __synth_change_variable_20
state f_permutation_.i : bv<23>
  [next] ite(%1867, 23'x000000, concat(f_permutation_.i[21:0], %1859))
state f_permutation_.calc : bv<1>
  [next] ite(%1867, 0, or(and(f_permutation_.calc, not(%556)), %1859))
state __synth_variable_47 : bv<1>
  [next] __synth_variable_47
state __synth_variable_48 : bv<1>
  [next] __synth_variable_48
state __synth_variable_49 : bv<1>
  [next] __synth_variable_49
state __synth_variable_50 : bv<1>
  [next] __synth_variable_50
state __synth_variable_51 : bv<1>
  [next] __synth_variable_51
state __synth_variable_52 : bv<1>
  [next] __synth_variable_52
state __synth_variable_53 : bv<1>
  [next] __synth_variable_53
state __synth_variable_54 : bv<1>
  [next] __synth_variable_54
state __synth_change_variable_6 : bv<1>
  [next] __synth_change_variable_6
state __synth_variable_55 : bv<1>
  [next] __synth_variable_55
state __synth_variable_56 : bv<1>
  [next] __synth_variable_56
state __synth_variable_57 : bv<1>
  [next] __synth_variable_57
state __synth_variable_58 : bv<1>
  [next] __synth_variable_58
state __synth_variable_59 : bv<1>
  [next] __synth_variable_59
state __synth_variable_60 : bv<1>
  [next] __synth_variable_60
state __synth_variable_61 : bv<1>
  [next] __synth_variable_61
state __synth_variable_62 : bv<1>
  [next] __synth_variable_62
state __synth_change_variable_7 : bv<1>
  [next] __synth_change_variable_7
state __synth_variable_63 : bv<1>
  [next] __synth_variable_63
state __synth_change_variable_8 : bv<1>
  [next] __synth_change_variable_8
state __synth_variable_64 : bv<1>
  [next] __synth_variable_64
state __synth_variable_65 : bv<1>
  [next] __synth_variable_65
state __synth_variable_66 : bv<1>
  [next] __synth_variable_66
state __synth_variable_67 : bv<1>
  [next] __synth_variable_67
state __synth_variable_68 : bv<1>
  [next] __synth_variable_68
state __synth_variable_69 : bv<1>
  [next] __synth_variable_69
state __synth_variable_70 : bv<1>
  [next] __synth_variable_70
state __synth_variable_71 : bv<1>
  [next] __synth_variable_71
state __synth_change_variable_9 : bv<1>
  [next] __synth_change_variable_9
state __synth_variable_72 : bv<1>
  [next] __synth_variable_72
state __synth_variable_73 : bv<1>
  [next] __synth_variable_73
state __synth_variable_74 : bv<1>
  [next] __synth_variable_74
state __synth_variable_75 : bv<1>
  [next] __synth_variable_75
state __synth_variable_76 : bv<1>
  [next] __synth_variable_76
state __synth_variable_77 : bv<1>
  [next] __synth_variable_77
state __synth_variable_78 : bv<1>
  [next] __synth_variable_78
state __synth_variable_79 : bv<1>
  [next] __synth_variable_79
state __synth_change_variable_10 : bv<1>
  [next] __synth_change_variable_10
state __synth_variable_80 : bv<1>
  [next] __synth_variable_80
state __synth_variable_81 : bv<1>
  [next] __synth_variable_81
state __synth_variable_82 : bv<1>
  [next] __synth_variable_82
state __synth_variable_83 : bv<1>
  [next] __synth_variable_83
state __synth_variable_84 : bv<1>
  [next] __synth_variable_84
state __synth_variable_85 : bv<1>
  [next] __synth_variable_85
state __synth_variable_86 : bv<1>
  [next] __synth_variable_86
state __synth_variable_87 : bv<1>
  [next] __synth_variable_87
state __synth_change_variable_11 : bv<1>
  [next] __synth_change_variable_11
state __synth_change_variable_12 : bv<1>
  [next] __synth_change_variable_12
state __synth_variable_96 : bv<1>
  [next] __synth_variable_96
state __synth_variable_97 : bv<1>
  [next] __synth_variable_97
state __synth_variable_98 : bv<1>
  [next] __synth_variable_98
state __synth_variable_99 : bv<1>
  [next] __synth_variable_99
state __synth_variable_100 : bv<1>
  [next] __synth_variable_100
state __synth_variable_101 : bv<1>
  [next] __synth_variable_101
state __synth_variable_102 : bv<1>
  [next] __synth_variable_102
state __synth_variable_103 : bv<1>
  [next] __synth_variable_103
state __synth_change_variable_14 : bv<1>
  [next] __synth_change_variable_14
state padder_.state : bv<1>
  [next] ite(%1840, 0, ite(ite(__synth_change_variable_10, ite(__synth_variable_79, 0, ite(__synth_variable_78, 0, ite(__synth_variable_77, 0, ite(__synth_variable_76, state, ite(__synth_variable_75, out_ready_0, ite(__synth_variable_74, 0, ite(__synth_variable_73, in_ready, ite(__synth_variable_72, reset, clk)))))))), is_last), 1'b1, padder_.state))
state padder_.done : bv<1>
  [next] ite(%1840, 0, ite(and(padder_.state, %1818), 1'b1, padder_.done))
state padder_.i : bv<18>
  [next] ite(%1840, 18'x00000, ite(or(%1795, and(or(and(and(not(padder_.state), ite(__synth_change_variable_9, ite(__synth_variable_71, 0, ite(__synth_variable_70, 0, ite(__synth_variable_69, 0, ite(__synth_variable_68, state, ite(__synth_variable_67, out_ready_0, ite(__synth_variable_66, 0, ite(__synth_variable_65, is_last, ite(__synth_variable_64, reset, clk)))))))), in_ready)), %1819), and(padder_.state, %1819)), not(padder_.done))), and(concat(padder_.i[16:0], 1'b1), concat(%1796, concat(%1796, concat(%1796, concat(%1796, concat(%1796, concat(%1796, concat(%1796, concat(%1796, concat(%1796, concat(%1796, concat(%1796, concat(%1796, concat(%1796, concat(%1796, concat(%1796, concat(%1796, concat(%1796, %1796)))))))))))))))))), padder_.i))
state __synth_variable_95 : bv<1>
  [next] __synth_variable_95
state __synth_variable_94 : bv<1>
  [next] __synth_variable_94
state __synth_variable_93 : bv<1>
  [next] __synth_variable_93
state __synth_variable_92 : bv<1>
  [next] __synth_variable_92
state __synth_variable_91 : bv<1>
  [next] __synth_variable_91
state __synth_variable_90 : bv<1>
  [next] __synth_variable_90
state __synth_variable_89 : bv<1>
  [next] __synth_variable_89
state __synth_variable_88 : bv<1>
  [next] __synth_variable_88
state __synth_variable_46 : bv<1>
  [next] __synth_variable_46
state __synth_variable_45 : bv<1>
  [next] __synth_variable_45
state __synth_variable_44 : bv<1>
  [next] __synth_variable_44
state __synth_variable_43 : bv<1>
  [next] __synth_variable_43
state __synth_variable_42 : bv<1>
  [next] __synth_variable_42
state __synth_variable_41 : bv<1>
  [next] __synth_variable_41
state __synth_variable_40 : bv<1>
  [next] __synth_variable_40
state __synth_variable_39 : bv<1>
  [next] __synth_variable_39
state __synth_variable_38 : bv<1>
  [next] __synth_variable_38
state __synth_variable_37 : bv<1>
  [next] __synth_variable_37
state __synth_variable_36 : bv<1>
  [next] __synth_variable_36
state __synth_variable_35 : bv<1>
  [next] __synth_variable_35
state __synth_variable_34 : bv<1>
  [next] __synth_variable_34
state __synth_variable_33 : bv<1>
  [next] __synth_variable_33
state __synth_variable_32 : bv<1>
  [next] __synth_variable_32
state __synth_variable_31 : bv<1>
  [next] __synth_variable_31
state __synth_variable_30 : bv<1>
  [next] __synth_variable_30
state __synth_variable_29 : bv<1>
  [next] __synth_variable_29
state __synth_variable_28 : bv<1>
  [next] __synth_variable_28
state __synth_variable_27 : bv<1>
  [next] __synth_variable_27
state __synth_variable_26 : bv<1>
  [next] __synth_variable_26
state __synth_variable_25 : bv<1>
  [next] __synth_variable_25
state __synth_variable_24 : bv<1>
  [next] __synth_variable_24
state __synth_variable_23 : bv<1>
  [next] __synth_variable_23
state __synth_variable_22 : bv<1>
  [next] __synth_variable_22
state __synth_variable_21 : bv<1>
  [next] __synth_variable_21
state __synth_variable_20 : bv<1>
  [next] __synth_variable_20
state __synth_variable_19 : bv<1>
  [next] __synth_variable_19
state __synth_variable_18 : bv<1>
  [next] __synth_variable_18
state __synth_variable_17 : bv<1>
  [next] __synth_variable_17
state __synth_variable_16 : bv<1>
  [next] __synth_variable_16
state __synth_variable_15 : bv<1>
  [next] __synth_variable_15
state __synth_variable_14 : bv<1>
  [next] __synth_variable_14
state __synth_variable_13 : bv<1>
  [next] __synth_variable_13
state __synth_variable_12 : bv<1>
  [next] __synth_variable_12
state __synth_variable_11 : bv<1>
  [next] __synth_variable_11
state __synth_variable_10 : bv<1>
  [next] __synth_variable_10
state __synth_variable_9 : bv<1>
  [next] __synth_variable_9
state __synth_variable_8 : bv<1>
  [next] __synth_variable_8
state __synth_variable_7 : bv<1>
  [next] __synth_variable_7
state __synth_variable_6 : bv<1>
  [next] __synth_variable_6
state __synth_variable_5 : bv<1>
  [next] __synth_variable_5
state __synth_variable_4 : bv<1>
  [next] __synth_variable_4
state __synth_variable_3 : bv<1>
  [next] __synth_variable_3
state __synth_variable_2 : bv<1>
  [next] __synth_variable_2
state __synth_variable_1 : bv<1>
  [next] __synth_variable_1
state __synth_variable_0 : bv<1>
  [next] __synth_variable_0
state __synth_change_variable_13 : bv<1>
  [next] __synth_change_variable_13
state __synth_change_variable_5 : bv<1>
  [next] __synth_change_variable_5
state __synth_change_variable_4 : bv<1>
  [next] __synth_change_variable_4
state __synth_change_variable_3 : bv<1>
  [next] __synth_change_variable_3
state __synth_change_variable_2 : bv<1>
  [next] __synth_change_variable_2
state __synth_change_variable_1 : bv<1>
  [next] __synth_change_variable_1
state __synth_change_variable_0 : bv<1>
  [next] __synth_change_variable_0

